project/DistributionPackage.scala

```diff
@@ -267,26 +269,36 @@ object DistributionPackage {
   ): Boolean = {
     import scala.collection.JavaConverters._
 
-    val enso = distributionRoot / "bin" / batName("enso")
-    log.info(s"Executing $enso ${args.mkString(" ")}")
-    val pb  = new java.lang.ProcessBuilder()
-    val all = new java.util.ArrayList[String]()
-    val disablePrivateCheck = {
-      val findRun = args.indexOf("--run")
-      if (findRun >= 0 && findRun + 1 < args.size) {
-        val whatToRun = args(findRun + 1)
+    val enso             = distributionRoot / "bin" / batName("enso")
+    val pb               = new java.lang.ProcessBuilder()
+    val all              = new java.util.ArrayList[String]()
+    val runArgumentIndex = locateRunArgument(args)
+    val runArgument      = runArgumentIndex.map(args)
+    val disablePrivateCheck = runArgument match {
+      case Some(whatToRun) =>
         if (whatToRun.startsWith("test/") && whatToRun.endsWith("_Tests")) {
           whatToRun.contains("_Internal_")
         } else {
           false
         }
-      } else {
-        false
-      }
+      case None => false
     }
-    all.add(enso.getAbsolutePath())
+
+    val runArgumentAsFile = runArgument.flatMap(createFileIfValidPath)
+    val projectDirectory  = runArgumentAsFile.flatMap(findProjectRoot)
+    val cwdOverride: Option[File] =
+      projectDirectory.flatMap(findParentFile).map(_.getAbsoluteFile)
+
+    all.add(enso.getAbsolutePath)
     all.addAll(args.asJava)
-    pb.command(all)
+    // Override the working directory of new process to be the parent of the project directory.
+    cwdOverride.foreach { c =>
+      pb.directory(c)
+    }
+    if (cwdOverride.isDefined) {
+      // If the working directory is changed, we need to translate the path - make it absolute.
+      all.set(runArgumentIndex.get + 1, runArgumentAsFile.get.getAbsolutePath)
+    }
```

Comment:
My only concern is that because the CWD changes, if there were any other paths given as arguments, they may now become invalid.

But I don't think we often add paths to `runEngineDistribution`, do we? Maybe I should add a comment to be sure to prefer absolute paths?

I was thinking of a 'heuristic' that will convent relative paths to absolute, but it would be too imprecise - is a `test` argument a path to the `./test/` directory or a completely different thing? Impossible to tell without context.

Reply:
I think the only path I really care about is after `--run`.